diff --git a/grub-core/kern/main.c b/grub-core/kern/main.c
index 3262444..c73eea6 100644
--- a/grub-core/kern/main.c
+++ b/grub-core/kern/main.c
@@ -212,7 +212,6 @@ grub_main (void)
 
   /* Hello.  */
   grub_setcolorstate (GRUB_TERM_COLOR_HIGHLIGHT);
-  grub_printf ("Welcome to GRUB!\n\n");
   grub_setcolorstate (GRUB_TERM_COLOR_STANDARD);
 
   /* Load pre-loaded modules and free the space.  */
diff --git a/grub-core/normal/menu.c b/grub-core/normal/menu.c
index 7e0a158..2018cb7 100644
--- a/grub-core/normal/menu.c
+++ b/grub-core/normal/menu.c
@@ -669,12 +669,10 @@ run_menu (grub_menu_t menu, int nested, int *auto_boot)
 
 /* Callback invoked immediately before a menu entry is executed.  */
 static void
-notify_booting (grub_menu_entry_t entry,
+notify_booting (grub_menu_entry_t entry __attribute__((unused)),
 		void *userdata __attribute__((unused)))
 {
-  grub_printf ("  ");
-  grub_printf_ (N_("Booting `%s'"), entry->title);
-  grub_printf ("\n\n");
+  grub_env_set ("color_normal", "black/black");
 }
 
 /* Callback invoked when a default menu entry executed because of a timeout
diff --git a/include/grub/term.h b/include/grub/term.h
index bf4dcb4..2021259 100644
--- a/include/grub/term.h
+++ b/include/grub/term.h
@@ -412,6 +412,8 @@ grub_term_setcursor (struct grub_term_output *term, int on)
 static inline void 
 grub_term_cls (struct grub_term_output *term)
 {
+  term->setcursor (term, 0);
+
   if (term->cls)
     (term->cls) (term);
   else
